src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -657,9 +484,359 @@ protected virtual void NotifyListeners(ProjectChangeEventArgs e)
         }
     }
 
+    internal override void UpdateProject(
+        string filePath,
+        RazorConfiguration configuration,
+        ProjectWorkspaceState projectWorkspaceState,
+        string? rootNamespace,
+        Func<IProjectSnapshot, ImmutableArray<IUpdateProjectAction>> calculate)
+    {
+        if (projectWorkspaceState is null)
+        {
+            throw new ArgumentNullException(nameof(projectWorkspaceState));
+        }
+
+        using var _ = ListPool<ProjectChangeEventArgs>.GetPooledObject(out var changesToNotify);
+
+        // Get an upgradeableLock, which will keep a read lock while we compute the changes
+        // and then get a write lock to actually apply them. Only one upgradeable lock
+        // can be held at any given time. Write lock must be retrieved on the same
+        // thread that the lock was acquired
+        using (var upgradeableLock = _locksFactory.EnterUpgradeAbleReadLock())
+        {
+            UpdateProject_NoLock(filePath, configuration, projectWorkspaceState, rootNamespace, calculate, changesToNotify, upgradeableLock);
+        }
+
+        // Notify outside of the lock, since notifications may trigger mutations from listeners
+        foreach (var notification in changesToNotify)
+        {
+            NotifyListeners(notification);
+        }
+    }
+
+    private void UpdateProject_NoLock(
+        string filePath,
+        RazorConfiguration configuration,
+        ProjectWorkspaceState projectWorkspaceState,
+        string? rootNamespace,
+        Func<IProjectSnapshot, ImmutableArray<IUpdateProjectAction>> calculate,
+        List<ProjectChangeEventArgs> changesToNotify,
+        in ReadWriterLocker.UpgradeableReadLock upgradeableLock)
```

Comment:
Bear in mind that `UpgradeableReadLock` is a mutable struct. So, if it mutates within this method (e.g. if it gets disposed), it might result in strange bugs because the caller won't see that mutation.